File: backend/app/services/purchase_service.py
import math
import re
import secrets
from datetime import datetime, timezone
from decimal import Decimal, ROUND_HALF_UP
from typing import Dict, Any, Optional, List
from bson import ObjectId
from pymongo.database import Database
from fastapi import HTTPException, status

from app.config import settings
from app.schemas.purchases import (
    CreatePurchaseRequest,
    PurchaseResponse,
    CustomerPurchaseListResponse,
    AdminPurchaseCustomerInfo,
    AdminPurchaseListItem,
    AdminPurchaseListResponse,
    AdminPurchaseDetailResponse,
)
from app.services.metal_rates_service import ensure_rates_initialized, _check_and_expire_rates
from app.services.holdings_service import process_purchase_for_holdings

_purchase_indexes_initialized = False


def _to_object_id(id_val: Any) -> Optional[ObjectId]:
    """Helper to safely convert string to ObjectId."""
    if isinstance(id_val, ObjectId):
        return id_val
    try:
        return ObjectId(str(id_val))
    except Exception:
        return None


def _to_decimal(val: Any) -> Decimal:
    """Safely convert any numeric value to Decimal."""
    return Decimal(str(val))


def _quantize_2(val: Decimal) -> float:
    """Format and round monetary decimal values cleanly to 2 decimal places."""
    return float(val.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP))


def _generate_transaction_id(metal: str) -> str:
    """Generate a unique, searchable, immutable transaction ID."""
    prefix = metal.upper()
    date_str = datetime.now(timezone.utc).strftime("%Y%m%d")
    random_hex = secrets.token_hex(4).upper()
    return f"{prefix}-{date_str}-{random_hex}"


def ensure_purchase_indexes(db: Database):
    """Ensure indexes for purchase collections."""
    global _purchase_indexes_initialized
    if _purchase_indexes_initialized or db is None:
        return
    try:
        db.purchases.create_index("user_id", name="idx_purchases_user_id")
        db.purchases.create_index("transaction_id", unique=True, name="uniq_purchases_transaction_id")
        db.purchases.create_index([("metal", 1), ("created_at", -1)], name="idx_purchases_metal_date")
        db.purchases.create_index("status", name="idx_purchases_status")
        db.purchases.create_index("payment_status", name="idx_purchases_payment_status")
        db.purchases.create_index("created_at", name="idx_purchases_created_at")
        _purchase_indexes_initialized = True
    except Exception:
        pass


def _format_purchase_response(doc: Dict[str, Any]) -> PurchaseResponse:
    """Format MongoDB purchase document into safe PurchaseResponse schema."""
    return PurchaseResponse(
        purchase_id=str(doc.get("_id", doc.get("id"))),
        transaction_id=doc.get("transaction_id", ""),
        metal=doc.get("metal", ""),
        quantity_grams=float(doc.get("quantity_grams", 0.0)),
        rate_per_gram=float(doc.get("rate_per_gram", 0.0)),
        metal_value=float(doc.get("metal_value", 0.0)),
        gst_rate_percent=float(doc.get("gst_rate_percent", 3.0)),
        gst_amount=float(doc.get("gst_amount", 0.0)),
        total_amount=float(doc.get("total_amount", 0.0)),
        currency=doc.get("currency", "INR"),
        status=doc.get("status", "completed"),
        payment_status=doc.get("payment_status", "paid"),
        created_at=doc.get("created_at"),
    )


def create_purchase(
    db: Database,
    current_user: Dict[str, Any],
    data: CreatePurchaseRequest,
) -> PurchaseResponse:
    """Create a new metal purchase calculating live rates and GST securely on the backend."""
    if db is None:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database service unavailable")

    ensure_purchase_indexes(db)
    ensure_rates_initialized(db)
    _check_and_expire_rates(db)

    metal = data.metal.lower()
    if metal not in ["gold", "silver"]:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Metal must be either 'gold' or 'silver'")

    # Validate minimum purchase weight
    if metal == "gold" and data.quantity_grams < settings.MIN_GOLD_PURCHASE_GRAMS:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Minimum gold purchase quantity is {settings.MIN_GOLD_PURCHASE_GRAMS} grams",
        )
    if metal == "silver" and data.quantity_grams < settings.MIN_SILVER_PURCHASE_GRAMS:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Minimum silver purchase quantity is {settings.MIN_SILVER_PURCHASE_GRAMS} grams",
        )

    # Fetch live operational rate for the requested metal
    rate_doc = db.rates.find_one({"metal": metal})
    if not rate_doc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=f"Active rate for {metal} is currently unavailable",
        )

    active_rate = rate_doc.get("active_rate", rate_doc.get("api_rate"))
    if not active_rate or active_rate <= 0:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=f"Active rate for {metal} is invalid",
        )

    # Precise financial calculations using Decimal
    qty_dec = _to_decimal(data.quantity_grams)
    rate_dec = _to_decimal(active_rate)
    gst_pct_dec = _to_decimal(settings.GST_RATE_PERCENT)

    metal_val_dec = qty_dec * rate_dec
    gst_amt_dec = metal_val_dec * (gst_pct_dec / Decimal("100"))
    total_amt_dec = metal_val_dec + gst_amt_dec

    metal_value = _quantize_2(metal_val_dec)
    gst_amount = _quantize_2(gst_amt_dec)
    total_amount = _quantize_2(total_amt_dec)
    rate_per_gram = _quantize_2(rate_dec)
    quantity_grams = float(qty_dec)

    user_obj_id = _to_object_id(current_user["id"])
    now_utc = datetime.now(timezone.utc)
    txn_id = _generate_transaction_id(metal)

    purchase_doc = {
        "user_id": user_obj_id if user_obj_id else current_user["id"],
        "metal": metal,
        "quantity_grams": quantity_grams,
        "rate_per_gram": rate_per_gram,
        "metal_value": metal_value,
        "gst_rate_percent": float(gst_pct_dec),
        "gst_amount": gst_amount,
        "total_amount": total_amount,
        "currency": "INR",
        "status": "completed",
        "payment_status": "paid",
        "transaction_id": txn_id,
        "created_at": now_utc,
        "updated_at": now_utc,
    }

    res = db.purchases.insert_one(purchase_doc)
    purchase_doc["_id"] = res.inserted_id

    # Automatically update customer holdings upon completed purchase
    try:
        process_purchase_for_holdings(db, purchase_doc)
    except Exception:
        pass

    return _format_purchase_response(purchase_doc)


def get_customer_purchases(
    db: Database,
    current_user: Dict[str, Any],
    metal: Optional[str] = None,
    status_filter: Optional[str] = None,
    page: int = 1,
    limit: int = 20,
) -> CustomerPurchaseListResponse:
    """Retrieve paginated purchases belonging strictly to the authenticated customer."""
    if db is None:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database service unavailable")

    user_obj_id = _to_object_id(current_user["id"])
    query: Dict[str, Any] = {
        "$or": [
            {"user_id": user_obj_id} if user_obj_id else {"user_id": current_user["id"]},
            {"user_id": current_user["id"]},
        ]
    }

    if metal and metal.lower().strip() in ["gold", "silver"]:
        query["metal"] = metal.lower().strip()

    if status_filter:
        query["status"] = status_filter.lower().strip()

    safe_limit = max(1, min(limit, 100))
    safe_page = max(1, page)
    skip = (safe_page - 1) * safe_limit

    total = db.purchases.count_documents(query)
    total_pages = max(1, math.ceil(total / safe_limit))

    cursor = db.purchases.find(query).sort("created_at", -1).skip(skip).limit(safe_limit)
    items = [_format_purchase_response(doc) for doc in cursor]

    return CustomerPurchaseListResponse(
        items=items,
        page=safe_page,
        limit=safe_limit,
        total=total,
        total_pages=total_pages,
    )


def get_customer_purchase_by_id(
    db: Database,
    current_user: Dict[str, Any],
    purchase_id: str,
) -> PurchaseResponse:
    """Retrieve a single purchase detail for the authenticated customer."""
    if db is None:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database service unavailable")

    purch_obj_id = _to_object_id(purchase_id)
    doc = None
    if purch_obj_id:
        doc = db.purchases.find_one({"_id": purch_obj_id})
    if not doc:
        doc = db.purchases.find_one({"transaction_id": purchase_id})
    if not doc:
        doc = db.purchases.find_one({"_id": purchase_id})

    if not doc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Purchase not found")

    # Security check: Ensure purchase belongs to current authenticated customer
    doc_user_id = str(doc.get("user_id"))
    req_user_id = str(current_user.get("id", current_user.get("_id")))
    if doc_user_id != req_user_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Purchase not found")

    return _format_purchase_response(doc)


def get_admin_purchases(
    db: Database,
    page: int = 1,
    limit: int = 20,
    search: Optional[str] = None,
    metal: Optional[str] = None,
    status_filter: Optional[str] = None,
    payment_status: Optional[str] = None,
) -> AdminPurchaseListResponse:
    """Retrieve paginated purchases for Admin portal with search and filtering."""
    if db is None:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database service unavailable")

    query: Dict[str, Any] = {}

    if metal and metal.lower().strip() in ["gold", "silver"]:
        query["metal"] = metal.lower().strip()

    if status_filter:
        query["status"] = status_filter.lower().strip()

    if payment_status:
        query["payment_status"] = payment_status.lower().strip()

    if search and search.strip():
        search_term = search.strip()
        search_pattern = re.escape(search_term)
        search_regex = {"$regex": search_pattern, "$options": "i"}

        # Search matching users
        matching_users = list(db.users.find({
            "$or": [
                {"name": search_regex},
                {"mobile": search_regex},
                {"email": search_regex},
            ]
        }, {"_id": 1}))
        matching_user_ids = [u["_id"] for u in matching_users]

        query["$or"] = [
            {"transaction_id": search_regex},
            {"user_id": {"$in": matching_user_ids}},
        ]

    safe_limit = max(1, min(limit, 100))
    safe_page = max(1, page)
    skip = (safe_page - 1) * safe_limit

    total = db.purchases.count_documents(query)
    total_pages = max(1, math.ceil(total / safe_limit))

    cursor = db.purchases.find(query).sort("created_at", -1).skip(skip).limit(safe_limit)

    items: List[AdminPurchaseListItem] = []
    for doc in cursor:
        user_id_val = doc.get("user_id")
        user_doc = db.users.find_one({"_id": user_id_val}) if user_id_val else None

        customer_info = AdminPurchaseCustomerInfo(
            user_id=str(user_id_val),
            name=user_doc.get("name", "") if user_doc else "Unknown",
            mobile=user_doc.get("mobile", "") if user_doc else "",
            email=user_doc.get("email") if user_doc else None,
        )

        items.append(
            AdminPurchaseListItem(
                purchase_id=str(doc["_id"]),
                transaction_id=doc.get("transaction_id", ""),
                customer=customer_info,
                metal=doc.get("metal", ""),
                quantity_grams=float(doc.get("quantity_grams", 0.0)),
                rate_per_gram=float(doc.get("rate_per_gram", 0.0)),
                total_amount=float(doc.get("total_amount", 0.0)),
                status=doc.get("status", "completed"),
                payment_status=doc.get("payment_status", "paid"),
                created_at=doc.get("created_at"),
            )
        )

    return AdminPurchaseListResponse(
        items=items,
        page=safe_page,
        limit=safe_limit,
        total=total,
        total_pages=total_pages,
    )


def get_admin_purchase_by_id(
    db: Database,
    purchase_id: str,
) -> AdminPurchaseDetailResponse:
    """Retrieve full purchase details and attached customer profile for Admin."""
    if db is None:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database service unavailable")

    purch_obj_id = _to_object_id(purchase_id)
    doc = None
    if purch_obj_id:
        doc = db.purchases.find_one({"_id": purch_obj_id})
    if not doc:
        doc = db.purchases.find_one({"transaction_id": purchase_id})
    if not doc:
        doc = db.purchases.find_one({"_id": purchase_id})

    if not doc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Purchase not found")

    user_id_val = doc.get("user_id")
    user_doc = db.users.find_one({"_id": user_id_val}) if user_id_val else None

    customer_info = AdminPurchaseCustomerInfo(
        user_id=str(user_id_val),
        name=user_doc.get("name", "") if user_doc else "Unknown",
        mobile=user_doc.get("mobile", "") if user_doc else "",
        email=user_doc.get("email") if user_doc else None,
    )

    return AdminPurchaseDetailResponse(
        purchase_id=str(doc["_id"]),
        transaction_id=doc.get("transaction_id", ""),
        customer=customer_info,
        metal=doc.get("metal", ""),
        quantity_grams=float(doc.get("quantity_grams", 0.0)),
        rate_per_gram=float(doc.get("rate_per_gram", 0.0)),
        metal_value=float(doc.get("metal_value", 0.0)),
        gst_rate_percent=float(doc.get("gst_rate_percent", 3.0)),
        gst_amount=float(doc.get("gst_amount", 0.0)),
        total_amount=float(doc.get("total_amount", 0.0)),
        currency=doc.get("currency", "INR"),
        status=doc.get("status", "completed"),
        payment_status=doc.get("payment_status", "paid"),
        created_at=doc.get("created_at"),
        updated_at=doc.get("updated_at"),
    )

File: backend/src/services/purchaseService.js
import { v4 as uuidv4 } from 'uuid';
import { query } from '../config/db.js';
import config from '../config/env.js';
import { cleanRate, cleanGrams, generateTransactionId } from '../utils/formatters.js';
import { getRatesPublic } from './metalRatesService.js';
import { processPurchaseForHoldings } from './holdingsService.js';
import { notifyPurchaseCompleted } from './notificationService.js';

export function formatPurchaseResponse(doc) {
  return {
    purchase_id: doc.id,
    transaction_id: doc.transaction_id,
    metal: doc.metal,
    quantity_grams: cleanGrams(doc.quantity_grams),
    rate_per_gram: cleanRate(doc.rate_per_gram),
    metal_value: cleanRate(doc.metal_value),
    gst_rate_percent: cleanRate(doc.gst_rate_percent || 3.0),
    gst_amount: cleanRate(doc.gst_amount),
    total_amount: cleanRate(doc.total_amount),
    currency: doc.currency || 'INR',
    status: doc.status || 'completed',
    payment_status: doc.payment_status || 'paid',
    created_at: doc.created_at,
  };
}

export async function createPurchase(user, data) {
  const metal = (data.metal || '').toLowerCase().trim();
  if (!['gold', 'silver'].includes(metal)) {
    const error = new Error("Metal must be either 'gold' or 'silver'");
    error.status = 400;
    throw error;
  }

  const quantityGrams = cleanGrams(data.quantity_grams);

  if (metal === 'gold' && quantityGrams < config.minGoldPurchaseGrams) {
    const error = new Error(`Minimum gold purchase quantity is ${config.minGoldPurchaseGrams} grams`);
    error.status = 400;
    throw error;
  }
  if (metal === 'silver' && quantityGrams < config.minSilverPurchaseGrams) {
    const error = new Error(`Minimum silver purchase quantity is ${config.minSilverPurchaseGrams} grams`);
    error.status = 400;
    throw error;
  }

  // Fetch active rate
  const rates = await getRatesPublic();
  const activeRate = cleanRate(rates[metal]?.active_rate);

  if (!activeRate || activeRate <= 0) {
    const error = new Error(`Active rate for ${metal} is currently unavailable`);
    error.status = 503;
    throw error;
  }

  // Financial calculations
  const metalValue = cleanRate(quantityGrams * activeRate);
  const gstAmount = cleanRate(metalValue * (config.gstRatePercent / 100));
  const totalAmount = cleanRate(metalValue + gstAmount);
  const transactionId = generateTransactionId(metal);

  const purchaseId = uuidv4();
  await query(
    `INSERT INTO purchases (id, transaction_id, user_id, metal, quantity_grams, rate_per_gram, metal_value, gst_rate_percent, gst_amount, total_amount, currency, status, payment_status, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'INR', 'completed', 'paid', NOW(), NOW())`,
    [
      purchaseId,
      transactionId,
      user.id,
      metal,
      quantityGrams,
      activeRate,
      metalValue,
      config.gstRatePercent,
      gstAmount,
      totalAmount,
    ]
  );

  const purchaseDoc = {
    id: purchaseId,
    transaction_id: transactionId,
    user_id: user.id,
    metal,
    quantity_grams: quantityGrams,
    rate_per_gram: activeRate,
    metal_value: metalValue,
    gst_rate_percent: config.gstRatePercent,
    gst_amount: gstAmount,
    total_amount: totalAmount,
    currency: 'INR',
    status: 'completed',
    payment_status: 'paid',
    created_at: new Date().toISOString(),
  };

  // Process holdings update
  try {
    await processPurchaseForHoldings(purchaseDoc);
  } catch (err) {
    console.error('Error processing holdings for purchase:', err);
  }

  // Trigger notification
  try {
    await notifyPurchaseCompleted(purchaseDoc);
  } catch (err) {
    console.error('Error sending purchase notification:', err);
  }

  return formatPurchaseResponse(purchaseDoc);
}

export async function getCustomerPurchases(user, { metal, status_filter, page = 1, limit = 20 }) {
  const safeLimit = Math.max(1, Math.min(parseInt(limit, 10) || 20, 100));
  const safePage = Math.max(1, parseInt(page, 10) || 1);
  const offset = (safePage - 1) * safeLimit;

  const whereClauses = ['user_id = ?'];
  const params = [user.id];

  if (metal && ['gold', 'silver'].includes(metal.toLowerCase().trim())) {
    whereClauses.push('metal = ?');
    params.push(metal.toLowerCase().trim());
  }

  if (status_filter) {
    whereClauses.push('status = ?');
    params.push(status_filter.toLowerCase().trim());
  }

  const whereSql = `WHERE ${whereClauses.join(' AND ')}`;
  const countRows = await query(`SELECT COUNT(*) as total FROM purchases ${whereSql}`, params);
  const total = countRows[0].total;
  const totalPages = Math.max(1, Math.ceil(total / safeLimit));

  const rows = await query(
    `SELECT * FROM purchases 
     ${whereSql} 
     ORDER BY created_at DESC 
     LIMIT ${safeLimit} OFFSET ${offset}`,
    params
  );

  const items = rows.map(formatPurchaseResponse);

  return {
    items,
    page: safePage,
    limit: safeLimit,
    total,
    total_pages: totalPages,
  };
}

export async function getCustomerPurchaseById(user, purchaseId) {
  const rows = await query(
    'SELECT * FROM purchases WHERE (id = ? OR transaction_id = ?) AND user_id = ? LIMIT 1',
    [purchaseId, purchaseId, user.id]
  );
  if (rows.length === 0) {
    const error = new Error('Purchase not found');
    error.status = 404;
    throw error;
  }
  return formatPurchaseResponse(rows[0]);
}

export async function getAdminPurchases({ page = 1, limit = 20, search, metal, status_filter, payment_status }) {
  const safeLimit = Math.max(1, Math.min(parseInt(limit, 10) || 20, 100));
  const safePage = Math.max(1, parseInt(page, 10) || 1);
  const offset = (safePage - 1) * safeLimit;

  const whereClauses = [];
  const params = [];

  if (metal && ['gold', 'silver'].includes(metal.toLowerCase().trim())) {
    whereClauses.push('p.metal = ?');
    params.push(metal.toLowerCase().trim());
  }

  if (status_filter) {
    whereClauses.push('p.status = ?');
    params.push(status_filter.toLowerCase().trim());
  }

  if (payment_status) {
    whereClauses.push('p.payment_status = ?');
    params.push(payment_status.toLowerCase().trim());
  }

  if (search && search.trim()) {
    const term = `%${search.trim()}%`;
    whereClauses.push('(p.transaction_id LIKE ? OR u.name LIKE ? OR u.mobile LIKE ? OR u.email LIKE ?)');
    params.push(term, term, term, term);
  }

  const whereSql = whereClauses.length > 0 ? `WHERE ${whereClauses.join(' AND ')}` : '';

  const countRows = await query(
    `SELECT COUNT(*) as total 
     FROM purchases p 
     JOIN users u ON p.user_id = u.id 
     ${whereSql}`,
    params
  );
  const total = countRows[0].total;
  const totalPages = Math.max(1, Math.ceil(total / safeLimit));

  const rows = await query(
    `SELECT p.*, u.name as customer_name, u.mobile as customer_mobile, u.email as customer_email
     FROM purchases p
     JOIN users u ON p.user_id = u.id
     ${whereSql}
     ORDER BY p.created_at DESC
     LIMIT ${safeLimit} OFFSET ${offset}`,
    params
  );

  const items = rows.map((r) => ({
    purchase_id: r.id,
    transaction_id: r.transaction_id,
    customer: {
      user_id: r.user_id,
      name: r.customer_name || 'Unknown',
      mobile: r.customer_mobile || '',
      email: r.customer_email || null,
    },
    metal: r.metal,
    quantity_grams: cleanGrams(r.quantity_grams),
    rate_per_gram: cleanRate(r.rate_per_gram),
    total_amount: cleanRate(r.total_amount),
    status: r.status,
    payment_status: r.payment_status,
    created_at: r.created_at,
  }));

  return {
    items,
    page: safePage,
    limit: safeLimit,
    total,
    total_pages: totalPages,
  };
}

export async function getAdminPurchaseById(purchaseId) {
  const rows = await query(
    `SELECT p.*, u.name as customer_name, u.mobile as customer_mobile, u.email as customer_email
     FROM purchases p
     JOIN users u ON p.user_id = u.id
     WHERE p.id = ? OR p.transaction_id = ?
     LIMIT 1`,
    [purchaseId, purchaseId]
  );

  if (rows.length === 0) {
    const error = new Error('Purchase not found');
    error.status = 404;
    throw error;
  }

  const r = rows[0];
  return {
    purchase_id: r.id,
    transaction_id: r.transaction_id,
    customer: {
      user_id: r.user_id,
      name: r.customer_name || 'Unknown',
      mobile: r.customer_mobile || '',
      email: r.customer_email || null,
    },
    metal: r.metal,
    quantity_grams: cleanGrams(r.quantity_grams),
    rate_per_gram: cleanRate(r.rate_per_gram),
    metal_value: cleanRate(r.metal_value),
    gst_rate_percent: cleanRate(r.gst_rate_percent || 3.0),
    gst_amount: cleanRate(r.gst_amount),
    total_amount: cleanRate(r.total_amount),
    currency: r.currency || 'INR',
    status: r.status,
    payment_status: r.payment_status,
    created_at: r.created_at,
    updated_at: r.updated_at,
  };
}

export default {
  createPurchase,
  getCustomerPurchases,
  getCustomerPurchaseById,
  getAdminPurchases,
  getAdminPurchaseById,
};
